File: src/main/scala/aoc/aoc2024/Day1.scala
package aoc.aoc2024

object Day1 extends aoc.Aoc("aoc2024/input1.txt", identity):
  val (a, b) = input.map{case s"$a   $b" => (a.toInt, b.toInt)}.unzip
  val part1 = a.sorted.zip(b.sorted).map((x, y) => (x - y).abs).sum
  println(part1)

  val s2 = b.groupMapReduce(identity)(_ => 1)(_ + _)
  val part2 = a.map(x => x * s2.getOrElse(x, 0)).sum
  println(part2)

File: src/main/kotlin/aoc/k2024/Day1.kt
package aoc.k2024

import aoc.getLines
import kotlin.math.absoluteValue

fun main() {
    val lines = getLines("/aoc2024/input1.txt")
    val (a, b) = lines
        .map { line -> line.split("   ").map { it.toInt() } }
        .map { Pair(it.first(), it.last()) }
        .unzip()
    val part1 = a.sorted().zip(b.sorted()).map{ (it.first - it.second).absoluteValue }.sum()
    println(part1)

    val s2 = b.groupingBy { it }.eachCount()
    val part2 = a.map { it * (s2[it] ?: 0) }.sum()
    println(part2)
}